relays/substrate/src/main.rs

```diff
@@ -517,3 +526,76 @@ async fn estimate_message_delivery_and_dispatch_fee<Fee: Decode, C: Chain, P: En
 		Decode::decode(&mut &encoded_response.0[..]).map_err(relay_substrate_client::Error::ResponseParseFailed)?;
 	Ok(decoded_response)
 }
+
+#[cfg(test)]
+mod tests {
+	use codec::Encode;
+	use sp_core::Pair;
+	use sp_runtime::traits::{IdentifyAccount, Verify};
+
+	macro_rules! call_for_runtime {
+		($runtime:tt) => {
+			$runtime::Call::System($runtime::SystemCall::remark(
+				std::format!(
+					"Unix time: {}",
+					std::time::SystemTime::now()
+						.duration_since(std::time::SystemTime::UNIX_EPOCH)
+						.unwrap_or_default()
+						.as_secs(),
+				)
+				.as_bytes()
+				.to_vec()
+			))
+		}
+	}
+
+	#[test]
+	fn millau_signature_is_valid_on_rialto() {
+		let millau_sign = relay_millau_client::SigningParams::from_suri("//Dave", None).unwrap();
+
+		let call = call_for_runtime!(rialto_runtime);
+
+		let millau_public: bp_millau::AccountSigner = millau_sign.signer.public().clone().into();
+		let millau_account_id: bp_millau::AccountId = millau_public.into_account();
+
+		// The `CallOrigin::TargetAccount` private key ownership proof is a signature
+		// over this message.
+		let mut message = Vec::new();
+		call.encode_to(&mut message);
+		millau_account_id.encode_to(&mut message);
+		rialto_runtime::VERSION.spec_version.encode_to(&mut message);
+		bp_runtime::MILLAU_BRIDGE_INSTANCE.encode_to(&mut message);
+
```

Comment:
This logic is now repeated 3 times (runtime, relay and test), IMHO we should create a standalone function that is responsible either for just constructing the message or constructing & verifying the signature, so that it can be re-used between here and the runtime (assuming we don't want the relayer code to depend on the runtime directly).

The main reason for this is to make sure that if we change this in only one of the places (i.e. only in runtime, not relayer) we will get a failing test, so we prevent a situation where this get's broken in the future.

Reply:
I think, I don't understand what you mean by 'runtime' here. I can see, how a separate function in the call_dispatch pallet might help. But that would not make the solution dependent on the runtime but the call_dispatch pallet. So, what am I missing here?

Reply:
By `runtime` I meant "the code that eventually ends up being part of the runtime", so `call-dispatch` pallet falls into that category :)
Yeah, just exposing a function in `call-dispatch` pallet should be good enough, or additionally you can have `millau_runtime` or `rialto_runtime` expose the same function but with some parameters already partially-applied (like runtime version or bridge instance).

Reply:
Next try.